research/delf/delf/python/datasets/tuples_dataset.py

```diff
@@ -0,0 +1,356 @@
+
+    Args:
+      net: Model, network to be used for negative re-mining.
+
+    Returns:
+      avg_l2: Float, average negative L2-distance.
+    """
+    self._n = 0
+    global_features_utils.debug_and_log(
+            '>> Creating tuples for an epoch of {}-{}...'.format(self._name,
+                                                                 self._mode),
+            True)
+    global_features_utils.debug_and_log(">> Used network: ", True)
+    global_features_utils.debug_and_log(net.meta_repr(), True)
+
+    ## Selecting queries.
+    # Draw `qsize` random queries for the tuples.
+    idx_list = np.arange(len(self._qpool))
+    np.random.shuffle(idx_list)
+    idxs2qpool = idx_list[:self._qsize]
+    self._qidxs = [self._qpool[i] for i in idxs2qpool]
+
+    ## Selecting positive pairs.
+    # Positives examples are fixed for each query during the whole training
+    # process.
+    self._pidxs = [self._ppool[i] for i in idxs2qpool]
+
+    ## Selecting negative pairs.
+    # If nnum = 0 create dummy nidxs.
+    # Useful when only positives used for training.
+    if self._nnum == 0:
+      self._nidxs = [[] for _ in range(len(self._qidxs))]
+      return 0
+
+    # Draw poolsize random images for pool of negatives images.
+    idx_list = np.arange(len(self.images))
+    np.random.shuffle(idx_list)
+    idxs2images = idx_list[:self._poolsize]
+
+    global_features_utils.debug_and_log(
+            '>> Extracting descriptors for query images...', debug=True)
+
+    img_list = self._img_names_to_full_path([self.images[i] for i in
+                                             self._qidxs])
+    qvecs = extract_descriptors_from_image_paths(
+            net,
+            image_paths=img_list,
+            imsize=self._imsize,
+            print_freq=self._print_freq)
+
+    global_features_utils.debug_and_log(
+            '>> Extracting descriptors for negative pool...', debug=True)
+
+    poolvecs = extract_descriptors_from_image_paths(
+            net,
+            image_paths=self._img_names_to_full_path([self.images[i] for i in
+                                                      idxs2images]),
+            imsize=self._imsize,
+            print_freq=self._print_freq)
+
+    global_features_utils.debug_and_log('>> Searching for hard negatives...',
+                                        debug=True)
+
+    # Compute dot product scores and ranks.
+    scores = tf.linalg.matmul(poolvecs, qvecs, transpose_a=True)
+    ranks = tf.argsort(scores, axis=0, direction='DESCENDING')
+
+    avg_ndist = 0.
+    n_ndist = 0.
+
+    # Selection of negative examples.
+    self._nidxs = []
+    eps = 1e-6
+
+    for q, qidx in enumerate(self._qidxs):
+      # We are not using the query cluster, those images are potentially
+      # positive.
+      qcluster = self._clusters[qidx]
+      clusters = [qcluster]
+      nidxs = []
+      r = 0
+
+      while len(nidxs) < self._nnum:
```

Comment:
could this become an infinite loop? Can we add conditions to break? I guess that would be once we exhaust the number of images in idxs2images

Reply:
I added the ValueError, I believe it should not be possible now

Reply:
I see. But I think the code could still crash?
For example: if there are 2 clusters, 10 potential negative images, but num_negatives is 3. I guess the call to `potential = neg_images_idxs[ranks[rank, q]]` would crash when rank becomes 10?
The solution would be to break in case `rank` reaches the end index of `ranks`.

If I am missing something, please let me know :)